docs/Design/Alpha/Data Model.md

```diff
@@ -0,0 +1,215 @@
+        int price_cents "USD cent value for price to buy full Album"
+        string cover_art_uri 
+        string release_date "The release date of the Album (Nullable timestamp)"
+        int release_year "The release year of the Album (Nullable)"
+        string created_at "Timestamp"
+        string updated_at "Timestamp"
+    }
+    track {
+        int id PK 
+        string(31) slug UK "Unique human-readable URL-safe identifier"
+        string(255) name "The name of the Track"
+        string(16384) description "Long text description (Nullable)"
+        int track_number "The number which describes the Track's ordering within its Album"
+        int price_cents "USD cent value for price to buy individual Track (Nullable)"
+        boolean is_streamable "Allows Users to stream the Track for free"
+        boolean is_previewable "Allows Users to preview the Track before purchase"
+        int streams "The number of streams the Track has gotten"
+        int downloads "The number of downloads the Track has gotten"
+        string content_uri "URI from which to stream or download the Track"
+        int runtime_seconds "The length of the Track in seconds"
+        string content_type "The format (MIME-type) of the Track's audio file"
+        int size_bytes "The size in bytes of the Track's audio file"
+        int sample_rate "The sample rate of the Track's audio file"
+        int bit_depth "The bits per sample of the Track's audio file"
+        string(255) checksum_sha256 "The SHA-256 checksum of the Track's audio file"
+        string created_at "Timestamp"
+        string updated_at "Timestamp"
+    }
+    playlist {
+        int id PK
+        string(31) slug UK "Unique human-readable URL-safe identifier"
+        string(255) name "The name of the Playlist"
+        string(16384) description "Long text description (Nullable)"
+        string(255) cover_art_uri "Nullable"
+        int owner_id FK "User who owns this Playlist"
+        string created_at "Timestamp"
+        string updated_at "Timestamp"
+    }
+    playlist_track {
+        int id PK
+        int playlist_id FK
+        int track_id FK
+        int manual_order "The manual sorting order for the Track in the Playlist"
+        string added_at "Timestamp"
+        string updated_at "Timestamp"
+    } 
+    user {
+        int id PK 
+        string(31) username UK ""
+        string(255) email UK ""
+        string(31) phone "Nullable"
+        string created_at "Timestamp"
+        string updated_at "Timestamp"
+    }
+    user_purchase_album {
+        int id PK
+        int user_id FK
+        int album_id FK
+        int purchase_price "Nullable"
+        string(255) invoice_uri UK
+        string purchased_at "Timestamp"
+    }
+    user_purchase_track {
+        int id PK
+        int user_id FK
+        int track_id FK
+        int purchase_price "Nullable"
+        string(255) invoice_uri UK
+        string purchased_at "Timestamp"
+    }
```

Comment:
Would it be better to just have it be called `purchase`, and then have the type of purchase be either a track, album, merch, etc?

Reply:
In some ways yes, however, foreign keys  need to be directly tied to another table. The alternative option would be to just hold an integer instead of a foreign key, but then 1) there will be collision issues we'd need to resolve, and 2) the database will be unaware of the relationship between `purchases` and `track`/`album`/`merch`/etc. It is possible we reorganize this to be one table with different columns for each "purchasable" model, eg.

```mermaid
erDiagram
purchase {
       int id PK
        int user_id FK
        int track_id FK "Nullable"
        int album_id FK "Nullable"
        int purchase_price "Nullable"
        string(255) invoice_uri UK
        string purchased_at "Timestamp"
}
```

and have a custom constraint such that only one of those columns can be non-null. Thoughts?

Reply:
if the original idea would make it easier to work on in the back end, I'm all for it then.